File: csharp/TetrisGameBase/logic/board/Board.cs
using hu.klenium.tetris.logic.tetromino;
using hu.klenium.tetris.util;

namespace hu.klenium.tetris.logic.board
{
    public class Board
    {
        public bool[,] Grid { get; private set; }
        public Dimension Size
        {
            get { return new Dimension(Grid.GetLength(0), Grid.GetLength(1)); }
        }
        public Board(Dimension size)
        {
            this.Grid = new bool[size.width, size.height];
        }
        public void AddTetromino(Tetromino tetromino)
        {
            foreach (Point partOffset in tetromino.Parts)
            {
                (int x, int y) = tetromino.Position + partOffset;
                Grid[x, y] = true;
            }
        }
        public bool CanAddTetromino(Tetromino tetromino, Point from)
        {
            if (!IsBoxInsideGrid(from, tetromino.BoundingBox))
                return false;
            foreach (Point partOffset in tetromino.Parts)
            {
                (int x, int y) = from + partOffset;
                if (Grid[x, y])
                    return false;
            }
            return true;
        }
        public void RemoveFullRows(int startRow, int numOfRowsToCheck)
        {
            for (int currentRow = startRow; currentRow < (startRow + numOfRowsToCheck); ++currentRow)
            {
                if (IsRowFull(currentRow))
                    MoveDownBoardsUpperPart(currentRow);
            }
        }

        private bool IsBoxInsideGrid(Point boxPosition, Dimension boxSize)
        {
            return !(boxPosition < (0, 0))
                && !((boxPosition + boxSize) > Size);
        }
        private bool IsRowFull(int rowIndex)
        {
            for (int x = 0; x < Size.width; ++x)
            {
                if (!Grid[x, rowIndex])
                    return false;
            }
            return true;
        }
        private void MoveDownBoardsUpperPart(int startRowIndex)
        {
            for (int y = startRowIndex; y > 0; --y)
            {
                for (int x = 0; x < Size.width; ++x)
                    Grid[x, y] = Grid[x, y - 1];
            }
            for (int x = 0; x < Size.width; ++x)
                Grid[x, 0] = false;
        }
    }
}

File: kotlin/src/hu/klenium/tetris/logic/board/Board.kt
package hu.klenium.tetris.logic.board

import hu.klenium.tetris.logic.tetromino.Tetromino
import hu.klenium.tetris.util.Dimension
import hu.klenium.tetris.util.Point

class Board(size: Dimension) {
    var grid: Array<BooleanArray> = Array(size.width) { BooleanArray(size.height)}
        private set
    val size: Dimension
        get() = Dimension(grid.size, grid[0].size)

    fun addTetromino(tetromino: Tetromino) {
        for (partOffset in tetromino.parts) {
            val (x, y) = tetromino.position + partOffset
            grid[x][y] = true
        }
    }
    fun canAddTetromino(tetromino: Tetromino, from: Point): Boolean {
        if (!isBoxInsideGrid(from, tetromino.boundingBox))
            return false
        for (partOffset in tetromino.parts) {
            val (x, y) = from + partOffset
            if (grid[x][y])
                return false
        }
        return true
    }
    fun removeFullRows() {

    }

    private fun isBoxInsideGrid(boxPosition: Point, boxSize: Dimension): Boolean {
        return boxPosition.x >= 0
            && boxPosition.y >= 0
            && (boxPosition.x + boxSize.width) <= size.width
            && (boxPosition.y + boxSize.height) <= size.height
    }
    private fun isRowFull(rowIndex: Int) : Boolean {
        TODO("not yet implemented")
    }
    private fun moveDownBoardsUpperPart(startRowIndex: Int) {

    }
}